File: C++/Merge_Sort.cpp
#include<iostream>
#include<vector>

using namespace std;

void merge(vector<int> &arr,int start,int mid,int end){
    vector<int> v;

    int i=start;
    int j=mid+1;

    while(i<=mid && j<=end){
        if(arr[i]<arr[j]){
            v.push_back(arr[i]);
            i++;
        }
        else{
            v.push_back(arr[j]);
            j++;
        }
    }

    while(i<=mid){
        v.push_back(arr[i]);
        i++;
    }
    while(j<=end){
        v.push_back(arr[j]);
        j++;
    }

    for(int i=0;i<v.size();i++){
        arr[i+start]=v[i];
    }
}

void Merge_Sort(vector<int> &arr,int start,int end){

    if(start>=end){
        return;
    }

    int mid=start + (end-start)/2;

    Merge_Sort(arr,start,mid);
    Merge_Sort(arr,mid+1,end);

    merge(arr,start,mid,end);
}

int main(){
    vector<int> arr={1,4,2,1,2,3,5,9};

    Merge_Sort(arr,0,7);

    for(int i=0;i<8;i++){
        cout << arr[i] << " ";
    }

    return 0;
}

File: Java/Merge_Sort.java
// Time Complexity : O(nlogn)

// Based On Divide & Conquer Approach

public class Merge_Sort {

    public static void Conquer(int[] arr,int start,int mid,int end){
        int[] merged = new int[end - start + 1];
        int idx = 0;
        int idx_1 = start;
        int idx_2 = mid + 1;

        while(idx_1 <= mid && idx_2 <= end){
            if(arr[idx_1] <= arr[idx_2]){
                merged[idx++] = arr[idx_1++];
            }
            else{
                merged[idx++] = arr[idx_2++];
            }
        }

        while(idx_1 <= mid){
            merged[idx++] = arr[idx_1++];
        }

        while (idx_2 <= end) {
            merged[idx++] = arr[idx_2++];
        }

        for(int i=0 , j=start ; i<merged.length ; i++ , j++){
            arr[j] = merged[i];
        }
    }

    public static void Divide(int[] arr,int start,int end){
        if(start >= end){
            return;
        }

        int mid = start + (end - start)/2;
        Divide(arr,start,mid);
        Divide(arr,mid+1,end);
        Conquer(arr,start,mid,end);
    }

    public static void main(String[] args){
        int arr[] = {1,4,2,1,2,1,3,9};

        Divide(arr,0,arr.length-1);

        for(int i:arr){
            System.out.print(i + " ");
        }
    } 
}
